File: 1-working-with-list/p01_last.ml

let rec last l =
  match l with
  | [] -> None
  | [e] -> Some e
  |  hd :: tl -> last tl;;

last [1; 2; 3; 4];;
last [];;

let rec last' = function
  | [] -> None
  | [e] -> Some e
  |  _ :: tl -> last tl;;

last' [1; 2; 3; 4];;
last' [];;

File: 1-working-with-list/p01_last.scala

object Main {

  def main(args: Array[String]) {
    println(last(List("a", "b", "c", "d")))
    println(last(List()))
  }

  def last[T](l: List[T]): Option[T] = l match {
    case Nil => None
    case hd :: Nil => Some(hd)
    case hd :: rest => last(rest)
  }
}
